### PR TITLE
[SPARK-32062][SQL] Reset listenerRegistered in SparkSession

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Reset listenerRegistered when application end.

### Why are the changes needed?

Within a jvm, stop and create `SparkContext` multi times will cause the bug.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Add UT.

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @cloud-fan @vinooganesh

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala [hunk 1]
```diff
@@ -240,4 +240,18 @@ class SparkSessionBuilderSuite extends SparkFunSuite with BeforeAndAfterEach {
     assert(session.conf.get(GLOBAL_TEMP_DATABASE) === "globaltempdb-spark-31532-2")
     assert(session.conf.get(WAREHOUSE_PATH) === "SPARK-31532-db-2")
   }
+
+  test("SPARK-32062: reset listenerRegistered in SparkSession") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:244]: This test relies on other tests to create and stop spark context, which is fragile.

Can we explicitly create and stop a spark context in this test?

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala [hunk 2]
```diff
@@ -240,4 +240,20 @@ class SparkSessionBuilderSuite extends SparkFunSuite with BeforeAndAfterEach {
...
+    (1 to 2).foreach { i =>
+      val conf = new SparkConf()
+        .setMaster("local")
+        .setAppName(s"test-SPARK-32062-$i")
+      val context = new SparkContext(conf)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:249]: does this work? The test doesn't stop the spark context, and AFAIK we don't support having multiple spark context instance at the same time.

Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala:249]: missed it.

**Comment by cloud-fan** [comment]
retest this please

**Comment by vinooganesh** [comment]
@ulysses-you - quick question for you (mostly for my own knowledge), do people usually destroy and re-create a spark context in the lifetime of a JVM? I was actually thinking about including this, but assumed that most people would create the singleton context and then use session instances for these types of short lives operations.

**Comment by ulysses-you** [comment]
@vinooganesh 
Not usual, but in local mode it may happen.
e.g. in spark test case, we create and stop context multi times. If the bug exists, other test case will be affected.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
@cloud-fan thanks for merging !
BTW do we need to a backport for 3.0.1 ?

**Comment by cloud-fan** [comment]
It's not a bug that people can hit in practice, I think it's OK to leave 3.0.

**Pull request closed without merging**